src/System.Runtime.InteropServices/src/System/Runtime/InteropServices/StandardOleMarshalObject.cs

```diff
@@ -0,0 +1,113 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Win32;
+using System.Diagnostics;
+
+namespace System.Runtime.InteropServices
+{
+    public class StandardOleMarshalObject : MarshalByRefObject, IMarshal
+    {
+        private static readonly Guid Clsid_StdMarshal = new Guid("00000017-0000-0000-c000-000000000046");
+        
+        [UnmanagedFunctionPointer(CallingConvention.StdCall)]
+        private delegate int GetMarshalSizeMax_Delegate(IntPtr _this, ref Guid riid, IntPtr pv, int dwDestContext, IntPtr pvDestContext, int mshlflags, out int pSize);
+        
+        [UnmanagedFunctionPointer(CallingConvention.StdCall)]
+        private delegate int MarshalInterface_Delegate(IntPtr _this, IntPtr pStm, ref Guid riid, IntPtr pv, int dwDestContext, IntPtr pvDestContext, int mshlflags);
+
+        protected StandardOleMarshalObject() { }
+
+        private IntPtr GetStdMarshaler(ref Guid riid, int dwDestContext, int mshlflags)
+        {
+            IntPtr pStandardMarshal = IntPtr.Zero;
+
+            IntPtr pUnknown = Marshal.GetIUnknownForObject(this);
+            if (pUnknown != IntPtr.Zero)
+            {
+                try
+                {
+                    if (NativeMethods.S_OK == Interop.Ole32.CoGetStandardMarshal(ref riid, pUnknown, dwDestContext, IntPtr.Zero, mshlflags, out pStandardMarshal))
+                    {
+                        Debug.Assert(pStandardMarshal != null, "Failed to get marshaler for interface '" + riid.ToString() + "', CoGetStandardMarshal returned S_OK");
+                        return pStandardMarshal;
+                    }
+                }
+                finally
+                {
+                    Marshal.Release(pUnknown);
+                }
+            }
+            throw new InvalidOperationException(string.Format(SR.StandardOleMarshalObjectGetMarshalerFailed, riid.ToString()));
```

Comment:
Nit: This doesn't feel correct. These COM methods are marked as [PreserveSig] but in here we are going to throw exception. since it works in desktop, maybe we just keep it as it is.